filedescriptor/src/unix.rs

```diff
@@ -217,6 +270,33 @@ impl FileDescriptor {
         }
         Ok(())
     }
+
+    /// Attempt to duplicate the underlying handle from an object that is
+    /// representable as the system `RawFileDescriptor` type and assign it to
+    /// a destination file descriptor. It then returns a `FileDescriptor`
+    /// wrapped around the duplicate.  Since the duplication requires kernel
+    /// resources that may not be available, this is a potentially fallible operation.
+    /// The returned handle has a separate lifetime from the source, but
+    /// references the same object at the kernel level.
+    pub unsafe fn dup2<F: AsRawFileDescriptor>(f: &F, dest_fd: RawFd) -> anyhow::Result<Self> {
+        OwnedHandle::dup2_impl(f, dest_fd).map(|handle| Self { handle })
+    }
+
+    pub(crate) fn redirect_stdio_impl<F: AsRawFileDescriptor>(
+        f: &F,
+        stdio: StdioDescriptor,
+    ) -> anyhow::Result<Self> {
+        let std_descriptor = match stdio {
+            StdioDescriptor::Stdin => libc::STDIN_FILENO,
+            StdioDescriptor::Stdout => libc::STDOUT_FILENO,
+            StdioDescriptor::Stderr => libc::STDERR_FILENO,
+        };
+
+        let std_original = FileDescriptor::dup(std_descriptor)?;
+        unsafe { FileDescriptor::dup2(f.into_raw_fd(), std_descriptor) }?;
```

Comment:
This doesn't compile because all we know about `F` is that it is `AsRawFileDescriptor`.  That means that we can call `as_raw_file_descriptor` to borrow (but not take ownership of) the `f` parameter.

```suggestion
        unsafe { FileDescriptor::dup2(f.as_raw_file_descriptor(), std_descriptor) }?;
```_

Reply:
Actually, I think we just want to pass this through:

```suggestion
        unsafe { FileDescriptor::dup2(f, std_descriptor) }?;
```

Reply:
Done